lib/cli/src/commands/package/download.rs

```diff
@@ -0,0 +1,177 @@
+use std::path::PathBuf;
+
+use anyhow::Context;
+use futures::StreamExt;
+use tokio::io::AsyncWriteExt;
+use wasmer_registry::wasmer_env::WasmerEnv;
+use wasmer_wasix::runtime::resolver::PackageSpecifier;
+
+/// Download a package from the registry.
+#[derive(clap::Parser, Debug)]
+pub struct CmdPackageDownload {
+    #[clap(flatten)]
+    env: WasmerEnv,
+
+    /// Verify that the downloaded file is a valid package.
+    #[clap(long)]
+    validate: bool,
+
+    /// Path where the package file should be written to.
+    /// If not specified, the data will be written to stdout.
+    #[clap(short = 'o', long)]
+    out_path: PathBuf,
+
+    /// The package to download.
+    /// Can be:
+    /// * a pakage specifier: `namespace/package[@vesion]`
+    /// * a URL
+    package: PackageSpecifier,
+}
+
+impl CmdPackageDownload {
+    pub(crate) fn execute(&self) -> Result<(), anyhow::Error> {
+        let rt = tokio::runtime::Runtime::new()?;
+        rt.block_on(self.run())
+    }
+
+    async fn run(&self) -> Result<(), anyhow::Error> {
+        if let Some(parent) = self.out_path.parent() {
+            match parent.metadata() {
+                Ok(m) => {
+                    if !m.is_dir() {
+                        anyhow::bail!(
+                            "parent of output file is not a directory: '{}'",
+                            parent.display()
+                        );
+                    }
+                }
+                Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                    std::fs::create_dir_all(parent)
+                        .context("could not create parent directory of output file")?;
+                }
+                Err(err) => return Err(err.into()),
+            }
+        };
+
+        let (url, token) = match &self.package {
+            PackageSpecifier::Registry { full_name, version } => {
+                let mut url = self.env.registry_public_url()?;
+                let p = format!("/{full_name}@{version}");
+                url.set_path(&p);
+                (url, self.env.token())
+            }
+            PackageSpecifier::Url(url) => {
+                (url.clone(), self.env.get_token_opt().map(|x| x.to_string()))
+            }
+            PackageSpecifier::Path(_) => {
+                anyhow::bail!("cannot download a package from a local path");
+            }
+        };
+
+        let client = reqwest::Client::new();
+        let mut b = client
+            .get(url.clone())
+            .header(http::header::ACCEPT, "application/webc");
+        if let Some(token) = token {
+            b = b.header(http::header::AUTHORIZATION, format!("Bearer {token}"));
+        };
+
+        eprintln!("Downloading package from '{url}'...");
```

Comment:
Instead of printing progress messages directly to stderr, can you make a spinner using `indicatif::ProgressBar`?

Reply:
Not doing this now, can be a followup.